packages/vite/src/node/optimizer/scan.ts

```diff
@@ -153,82 +153,73 @@ export function scanImports(environment: ScanEnvironment): {
         .map((entry) => `\n  ${colors.dim(entry)}`)
         .join('')}`,
     )
-    return prepareEsbuildScanner(
-      environment,
-      entries,
-      deps,
-      missing,
-      scanContext,
-    )
-  })
+    const deps: Record<string, string> = {}
+    const missing: Record<string, string> = {}
+
+    let context: BuildContext | undefined
+    try {
+      esbuildContext = prepareEsbuildScanner(
+        environment,
+        entries,
+        deps,
+        missing,
+      )
+      context = await esbuildContext
+      if (scanContext.cancelled) return
 
-  const result = esbuildContext
-    .then((context) => {
-      function disposeContext() {
-        return context?.dispose().catch((e) => {
-          environment.logger.error('Failed to dispose esbuild context', {
-            error: e,
-          })
-        })
-      }
-      if (!context || scanContext.cancelled) {
-        disposeContext()
-        return { deps: {}, missing: {} }
-      }
-      return context
-        .rebuild()
-        .then(() => {
-          return {
-            // Ensure a fixed order so hashes are stable and improve logs
-            deps: orderedDependencies(deps),
-            missing,
-          }
-        })
-        .finally(() => {
-          return disposeContext()
-        })
-    })
-    .catch(async (e) => {
-      if (e.errors && e.message.includes('The build was canceled')) {
-        // esbuild logs an error when cancelling, but this is expected so
-        // return an empty result instead
-        return { deps: {}, missing: {} }
-      }
+      try {
+        await context!.rebuild()
+        return {
+          // Ensure a fixed order so hashes are stable and improve logs
+          deps: orderedDependencies(deps),
+          missing,
+        }
+      } catch (e) {
+        if (e.errors && e.message.includes('The build was canceled')) {
+          // esbuild logs an error when cancelling, but this is expected so
+          // return an empty result instead
+          return { deps: {}, missing: {} }
```

Comment:
Should this return `undefined` like in line 168? Now that there is a fallback at 222, maybe it is good to normalize it. Or if not, return `{ deps: {}, missing: {} }` everywhere instead of the fallback

Reply:
That makes sense. I've updated to return `undefined`.